src/acl/FilledChecklist.h

```diff
@@ -117,7 +114,7 @@ class ACLFilledChecklist: public ACLChecklist
     err_type requestErrorType;
 
 private:
-    ConnStateData * conn_;          /**< hack for ident and NTLM */
+    ConnStateData * conn_; ///< client-to-Squid connection manager (if any)
```

Comment:
My understanding is that a) state that needs to be shared outside the Job itself is to be stored in a separate `RefCounted` object, and b) Job actions that need to be triggered or responses to pass back to a Job should use `AsyncCall` with a `Pointer` (or `Subscription`). Why? because the `AsyncJob` may self-delete/deallocate at any time due to its own reasons.

Reply:
I do not insist on a change of the raw-pointer here. Just to keep the "hack" wording.

Reply:
> My understanding is that a) state that needs to be shared outside the Job itself is to be stored in a separate `RefCounted` object

Yes, RefCount is one way to do that, but not the only one. Depending on the state and sharing needs, other information sharing mechanisms may be appropriate (e.g., copying or cbdata). There are too many different use cases to cover everything with RefCount.

and b) Job actions that need to be triggered or responses to pass back to a Job should use `AsyncCall`

Yes, and to call a job, one needs to store a pointer[^1] to that job. This is why there cannot be a "should not retain pointers to jobs" guideline.

[^1]: That job pointer may be hidden inside a stored AsyncCall, but there are other possibilities as well.

> with a `Pointer` (or `Subscription`).

Yes (or other proper means). Again, there are too many different use cases to cover everything with one approach (or two).

> Why? because the `AsyncJob` may self-delete/deallocate at any time due to its own reasons.

Indeed. Moreover, a job might be destroyed for reasons outside job's direct control -- we already have examples of jobs killing other (unsuspecting) jobs.

Someday, this `conn_` hack (that violates several guidelines!) will be gone because the information we extract via `conn_` today will be available in MasterXaction and ALE.

Thank you for updating your review and clearing this PR!